pkgs/applications/networking/instant-messengers/zulip/default.nix

```diff
@@ -1,35 +1,70 @@
 { lib
-, fetchurl
-, appimageTools
+, fetchFromGitHub
+, buildNpmPackage
+, electron_32
+, makeDesktopItem
+, copyDesktopItems
 }:
 
-let
+buildNpmPackage rec {
   pname = "zulip";
   version = "5.11.1";
 
-  src = fetchurl {
-    url = "https://github.com/zulip/zulip-desktop/releases/download/v${version}/Zulip-${version}-x86_64.AppImage";
-    hash = "sha256-t5qBm5+kTdeRMvcHpNbS5mp184UG/IqgJrtj7Ntcbb0=";
-    name="${pname}-${version}.AppImage";
+  src = fetchFromGitHub {
+    owner = "zulip";
+    repo = "zulip-desktop";
+    rev = "v${version}";
+    hash = "sha256-ELuQ/K5QhtS4QiTR35J9VtYNe1qBrS56Ay6mtcGL+FI=";
   };
 
-  appimageContents = appimageTools.extractType2 {
-    inherit pname version src;
-  };
+  npmDepsHash = "sha256-13Rlqa7TC2JUq6q1b2U5X3EXpOJGZ62IeF163/mTo68=";
+
+  ELECTRON_SKIP_BINARY_DOWNLOAD = 1;
+
+  nativeBuildInputs = [
+    copyDesktopItems
+  ];
+
+  dontNpmBuild = true;
+  buildPhase = ''
+    runHook preBuild
+
+    npm run pack -- \
+      -c.electronDist=${electron_32}/libexec/electron \
+      -c.electronVersion=${electron_32.version}
 
-in appimageTools.wrapType2 {
-  inherit pname version src;
+    runHook postBuild
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p "$out/share/lib/zulip"
+    cp -r dist/*-unpacked/resources/app.asar* "$out/share/lib/zulip/"
 
-  runScript = "appimage-exec.sh -w ${appimageContents} -- \${NIXOS_OZONE_WL:+\${WAYLAND_DISPLAY:+--ozone-platform-hint=auto}}";
+    install -m 444 -D app/resources/zulip.png $out/share/icons/hicolor/512x512/apps/zulip.png
 
-  extraInstallCommands = ''
-    install -m 444 -D ${appimageContents}/zulip.desktop $out/share/applications/zulip.desktop
-    install -m 444 -D ${appimageContents}/usr/share/icons/hicolor/512x512/apps/zulip.png \
-      $out/share/icons/hicolor/512x512/apps/zulip.png
-    substituteInPlace $out/share/applications/zulip.desktop \
-      --replace-fail 'Exec=AppRun' 'Exec=${pname}'
+    makeWrapper '${electron_32}/bin/electron' "$out/bin/zulip" \
```

Comment:
This should be `makeShellWrapper`, in case `makeBinaryWrapper` ever becomes the default, since the `${…:+…}` construction is shell-specific.

And `makeShellWrapper` should be added to `nativeBuildInputs` if we’re using it directly, even though it’s indirectly pulled in via `buildNpmPackage`.